### ISSUE TITLE
[Bug] Test set is taken from training set

### ISSUE THREAD

**Issue by Peter-Devine** [artidoro/qlora#251]
Hey, first of all, thanks for the great repo. This is great work.

I think I have found a bug regarding how the data is split in qlora.py

When you load a dataset from a local file, the dataset is [automatically split into train:test splits at a 90:10 ratio](https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L547). This results in a datasets dict with the keys `"train"` and `"test"`.

However, [the code later only checks whether a dataset has an `"eval"` key rather than a `"test"` key in the dict](https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L641C11-L641C11).

If the code doesn't have an `"eval"` key, then [a test set is generated from the `"train"` set](https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L645C1-L648C43), which is then [the full `"train"` set for training the model](https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L654). This would mean that we are evaluating using training data - not a good way to test whether the model has overfit or not.

A small change would account for this, by adding a similar check for a `"test"` key after [the check for an `"eval"` key](https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L641).

Thanks!

**Comment by Peter-Devine** [comment]
PS, being able to change the ratio (or number of examples) of the test set [when it is automatically generated](https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L547) (i.e., making it so that instead of a setting test_size as a constant with `test_size=0.1`, it would be an input variable `test_size=args.local_test_set_size`) would be a nice touch. Thanks again!
